nav-app/src/app/techniques-search/techniques-search.component.ts

```diff
@@ -37,71 +39,140 @@ export class TechniquesSearchComponent implements OnInit {
         }
     ]
 
+    private previousQuery: string = "";
     private _query: string = "";
     public set query(newQuery: string) {
-        let self = this;
+        this.previousQuery = this._query;
         this._query = newQuery;
-        if (this._query.trim() != "") {
-            // search
+        this.getStixResults(this._query);
+        this.getTechniqueResults(this._query);
+    }
 
-            //get master list of techniques and sub-techniques
-            let allTechniques = this.dataService.getDomain(this.viewModel.domainID).techniques;
-            for (let technique of allTechniques) {
-                allTechniques = allTechniques.concat(technique.subtechniques);
-            }
+    public get queryLength(): number {
+        return this._query.length;
+    }
+
+    public techniqueResults: Technique[] = [];
 
-            let results = allTechniques.filter(function(technique: Technique) {
+    constructor(private dataService: DataService) {
+        this.stixTypes = [];
+    }
+
+    ngOnInit() {
+        this.getStixData();
+        this.getTechniques();
+    }
+
+    // filterAndSort() takes an array of items and does the following:
+    // 1) if the query is empty, then it sorts the array
+    // 2) if the query is not empty, then it filters the already sorted array until nothing is left, or,
+    // the query is cleared out and empty again
+    filterAndSort(items: any[], query = "", sortAllTechniques = false) {
```

Comment:
Args should probably be documented in doc comment, it's not inherently apparent what sortAllTechniques does without a doc comment.

Reply:
Should `items` be `BaseStix[]` instead of `any[]`? It seems like it will always be a type which inherits from [BaseStix](https://github.com/mitre-attack/attack-navigator/blob/master/nav-app/src/app/data.service.ts#L308).